### ISSUE TITLE
Mistaken 'Name already defined' for decorated function

### ISSUE THREAD

**Issue by Zac-HD** [python/mypy#4117]
While working on HypothesisWorks/hypothesis-python#858, I've found that decorated functions are mis-detected as already defined in another branch, if either both functions or only the second are decorated.  I've simplified this to the following example:

```python
def decor(func):
    return func

if True: # or False, or any other condition
    def f(arg):
        yield
else:
    @decor  # not detected as duplicate if this decorator is removed
    def f(arg):
        yield
```

```console
$ python -m mypy example.py
a.py:8: error: Name 'f' already defined
```

This bug is still present as of mypy 0.610

**Comment by gvanrossum** [comment]
This check is trying to find true duplicates like
```py
def f(): pass
def f(): pass
```
but I agree when there are different definitions in two branches it's annoying. However, what would you want to happen if the two definitions are not compatible?

**Comment by Zac-HD** [comment]
The check actually handles a function with the same name being defined in different branches of a conditional - the problem only occurs if the second definition is decorated.  I've added a comment above to reflect this.  The full motivating example is this:

https://github.com/HypothesisWorks/hypothesis-python/blob/96800edf2714c4baed326f8eed450ea657a9fdef/src/hypothesis/internal/coverage.py#L98-L116

**Comment by Zac-HD** [comment]
The check actually handles a function with the same name being defined in different branches of a conditional - the problem only occurs if the second definition is decorated.  I've added a comment above to reflect this.  The full motivating example is [this code](https://github.com/HypothesisWorks/hypothesis-python/blob/96800edf2714c4baed326f8eed450ea657a9fdef/src/hypothesis/internal/coverage.py#L98-L116).

So "all that's needed"  for compatible functions is handling decorated functions in the existing heuristic.  

There are reasonable use-cases for differently-typed functions defined in branches (eg Py2/3 compatible str/bytes handling), so the 'correct enough' behavior would be to take the union of these function types where more than one could be in scope.  Currently, Mypy reveals the type of `f` to be whatever it was first defined as, and ignores later functions - should I open a new issue for this as a separate bug?

**Comment by gvanrossum** [comment]
I guess we should keep this bug for the case where at least one of the branches is decorated (assuming the decorator doesn't change the type).

I'm not sure what should be done (regardless decorators) if the types don't match -- in some cases you'd like this to infer the intersection of the two types, but if the duplication is accidental that could cause really confusing error messages for code elsewhere that's *calling* the function. In any case I believe we already have some issues about that. (If you feel the need to further discuss this, please search for relevant issues first and cross-link if you find one.)

**Comment by berislavlopac** [comment]
I have a problem with this error popping up on redefining decorated functions for `functools.singledispatch`.

**Comment by ethanhs** [comment]
@berislavlopac there is some discussion of what to do with the case of `singledispatch` here: https://github.com/python/mypy/issues/2904. At the moment a plugin handling the issues with `singledispatch` seem to the path forward.

**Comment by Zac-HD** [comment]
Seems to be fixed as of 0.761 :smile:
